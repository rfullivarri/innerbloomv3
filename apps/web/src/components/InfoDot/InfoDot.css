:root {
  --info-dot-size: 19.2px;
  --info-dot-font: 'Inter', system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif;
  --info-dot-bg: #151a33;
  --info-dot-border: rgba(122, 134, 255, 0.55);
  --info-dot-color: #e8eaff;
  --info-dot-hover-bg: #7d3cff;
  --info-dot-hover-color: #ffffff;
  --info-dot-shadow: 0 10px 24px rgba(70, 40, 130, 0.35);
  --info-pop-z: 9999;
  --info-pop-bg: #0f1328ee;
  --info-pop-border: rgba(122, 134, 255, 0.32);
  --info-pop-color: #ecedff;
  --info-pop-radius: 14px;
  --info-pop-padding-y: 14px;
  --info-pop-padding-x: 16px;
  --info-pop-font-size: 13px;
  --info-pop-line-height: 1.5;
  --info-pop-shadow: 0 18px 40px rgba(0, 0, 0, 0.55), inset 0 1px 0 rgba(255, 255, 255, 0.04);
  --info-pop-backdrop: blur(16px);
  --info-pop-arrow: 12px;
  --info-pop-spacing: 14px;
}

.info-dot-target {
  position: relative;
  display: inline-flex;
  align-items: center;
  gap: 0.5rem;
}

.info-dot-target > .info-dot__mount {
  display: inline-flex;
  align-items: center;
}

.info-dot-target[data-info-placement='left'] > .info-dot__mount {
  margin-right: 0.5rem;
}

.info-dot-target:not([data-info-placement='left']) > .info-dot__mount {
  margin-left: 0.5rem;
}

.info-dot {
  position: relative;
  display: inline-flex;
}

.info-dot__button {
  width: var(--info-dot-size);
  height: var(--info-dot-size);
  border-radius: 9999px;
  border: 1px solid var(--info-dot-border);
  background: var(--info-dot-bg);
  color: var(--info-dot-color);
  font-family: var(--info-dot-font);
  font-weight: 700;
  font-size: 15px;
  line-height: 1;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  transition: transform 120ms ease, background-color 120ms ease, color 120ms ease, box-shadow 120ms ease;
  box-shadow: var(--info-dot-shadow);
  touch-action: manipulation;
}

@media (hover: hover) {
  .info-dot__button:hover {
    transform: scale(1.06);
    background: var(--info-dot-hover-bg);
    color: var(--info-dot-hover-color);
  }
}

.info-dot__button:active {
  transform: scale(0.95);
}

.info-dot__button:focus-visible {
  outline: 2px solid rgba(199, 181, 255, 0.9);
  outline-offset: 3px;
}

.info-dot__popover {
  position: fixed;
  top: 0;
  left: 0;
  display: block;
  max-width: min(340px, 88vw);
  background: var(--info-pop-bg);
  border: 1px solid var(--info-pop-border);
  color: var(--info-pop-color);
  padding: var(--info-pop-padding-y) var(--info-pop-padding-x);
  border-radius: var(--info-pop-radius);
  box-shadow: var(--info-pop-shadow);
  backdrop-filter: var(--info-pop-backdrop);
  font-size: var(--info-pop-font-size);
  line-height: var(--info-pop-line-height);
  z-index: var(--info-pop-z);
  pointer-events: auto;
  animation: info-dot-fade-in 140ms ease-out;
}

.info-dot__popover:focus {
  outline: none;
}

.info-dot__popover::after {
  content: '';
  position: absolute;
  width: var(--info-pop-arrow);
  height: var(--info-pop-arrow);
  background: var(--info-pop-bg);
  box-shadow: 0 0 0 1px var(--info-pop-border);
  transform: rotate(45deg);
}

.info-dot__popover[data-placement='top']::after {
  bottom: calc(var(--info-pop-arrow) / -2);
  left: calc(var(--arrow-x, 50%) - var(--info-pop-arrow) / 2);
}

.info-dot__popover[data-placement='bottom']::after {
  top: calc(var(--info-pop-arrow) / -2);
  left: calc(var(--arrow-x, 50%) - var(--info-pop-arrow) / 2);
}

.info-dot__popover[data-placement='left']::after {
  right: calc(var(--info-pop-arrow) / -2);
  top: calc(var(--arrow-y, 50%) - var(--info-pop-arrow) / 2);
}

.info-dot__popover[data-placement='right']::after {
  left: calc(var(--info-pop-arrow) / -2);
  top: calc(var(--arrow-y, 50%) - var(--info-pop-arrow) / 2);
}

.info-dot__title {
  margin: 0 0 8px;
  font-size: 0.95rem;
  font-weight: 600;
  color: #f8f9ff;
}

.info-dot__list {
  margin: 0;
  padding-left: 1.1em;
  display: flex;
  flex-direction: column;
  gap: 0.35rem;
}

.info-dot__list li {
  margin: 0;
}

@keyframes info-dot-fade-in {
  from {
    opacity: 0;
    transform: translateY(4px) scale(0.96);
  }
  to {
    opacity: 1;
    transform: translateY(0) scale(1);
  }
}

@media (max-width: 480px) {
  :root {
    --info-dot-size: 21.6px;
  }

  .info-dot__popover {
    max-width: min(360px, 92vw);
    font-size: 13.5px;
  }
}
