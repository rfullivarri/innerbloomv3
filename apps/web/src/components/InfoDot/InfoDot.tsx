import { Fragment, useCallback, useEffect, useId, useLayoutEffect, useMemo, useRef, useState } from 'react';
import type { CSSProperties, ReactNode } from 'react';
import { createPortal } from 'react-dom';
import { infoTips, type InfoKey } from '../../content/infoTips';
import './InfoDot.css';

export type InfoDotPlacement = 'top' | 'right' | 'bottom' | 'left';

interface InfoDotProps {
  id: InfoKey;
  placement?: InfoDotPlacement;
  className?: string;
}

interface PositionState {
  top: number;
  left: number;
  placement: InfoDotPlacement;
  arrowX?: number;
  arrowY?: number;
}

const VIEWPORT_PADDING = 12;
const POPOVER_GAP = 18;
const MIN_ARROW_OFFSET = 12;

const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect;

function clamp(value: number, min: number, max: number) {
  if (Number.isNaN(value)) {
    return min;
  }
  return Math.min(Math.max(value, min), max);
}

function renderInline(text: string): ReactNode[] {
  return text
    .split(/(\*\*[^*]+\*\*)/g)
    .filter(Boolean)
    .map((segment, index) => {
      if (segment.startsWith('**') && segment.endsWith('**')) {
        return (
          <Fragment key={index}>
            <strong>{segment.slice(2, -2)}</strong>
          </Fragment>
        );
      }
      return <Fragment key={index}>{segment}</Fragment>;
    });
}

function combine(...classes: Array<string | false | null | undefined>) {
  return classes.filter(Boolean).join(' ');
}

function getPlacementOrder(desired: InfoDotPlacement): InfoDotPlacement[] {
  const base: InfoDotPlacement[] = ['top', 'bottom', 'right', 'left'];
  return [desired, ...base.filter((placement) => placement !== desired)];
}

function getFocusableElements(root: HTMLElement): HTMLElement[] {
  const selectors = [
    'a[href]',
    'button:not([disabled])',
    'input:not([disabled])',
    'select:not([disabled])',
    'textarea:not([disabled])',
    '[tabindex]:not([tabindex="-1"])',
  ];
  return Array.from(root.querySelectorAll<HTMLElement>(selectors.join(','))).filter(
    (element) => !element.hasAttribute('disabled') && !element.getAttribute('aria-hidden'),
  );
}

export function InfoDot({ id, placement = 'top', className }: InfoDotProps) {
  const tip = infoTips[id];
  if (!tip) {
    throw new Error(`InfoDot: no tip found for id "${id}"`);
  }

  const buttonRef = useRef<HTMLButtonElement>(null);
  const popoverRef = useRef<HTMLDivElement>(null);
  const shouldRestoreFocusRef = useRef(false);
  const pointerDownRef = useRef(false);
  const suppressFocusOpenRef = useRef(false);

  const [isMounted, setIsMounted] = useState(false);
  const [isOpen, setIsOpen] = useState(false);
  const [position, setPosition] = useState<PositionState>({
    top: 0,
    left: 0,
    placement,
  });


  const baseId = useId();
  const popoverId = `${baseId}-dialog`;
  const titleId = `${baseId}-title`;

  useEffect(() => {
    setIsMounted(true);
  }, []);

  const updatePosition = useCallback(() => {
    if (typeof window === 'undefined') {
      return;
    }
    const button = buttonRef.current;
    const popover = popoverRef.current;
    if (!button || !popover) {
      return;
    }

    const buttonRect = button.getBoundingClientRect();
    const popRect = popover.getBoundingClientRect();
    const viewportWidth = window.innerWidth;
    const viewportHeight = window.innerHeight;

    const width = Math.max(popRect.width || popover.offsetWidth || 1, 1);
    const height = Math.max(popRect.height || popover.offsetHeight || 1, 1);

    const buttonCenterX = buttonRect.left + buttonRect.width / 2;
    const buttonCenterY = buttonRect.top + buttonRect.height / 2;

    const candidates = getPlacementOrder(placement);
    let chosen: PositionState | null = null;
    let fallback: PositionState | null = null;

    for (const candidate of candidates) {
      let top = 0;
      let left = 0;
      let arrowX: number | undefined;
      let arrowY: number | undefined;
      let fits = true;

      switch (candidate) {
        case 'top': {
          top = buttonRect.top - height - POPOVER_GAP;
          left = buttonCenterX - width / 2;
          if (top < VIEWPORT_PADDING) {
            fits = false;
          }
          break;
        }
        case 'bottom': {
          top = buttonRect.bottom + POPOVER_GAP;
          left = buttonCenterX - width / 2;
          if (top + height > viewportHeight - VIEWPORT_PADDING) {
            fits = false;
          }
          break;
        }
        case 'left': {
          left = buttonRect.left - width - POPOVER_GAP;
          top = buttonCenterY - height / 2;
          if (left < VIEWPORT_PADDING) {
            fits = false;
          }
          break;
        }
        case 'right':
        default: {
          left = buttonRect.right + POPOVER_GAP;
          top = buttonCenterY - height / 2;
          if (left + width > viewportWidth - VIEWPORT_PADDING) {
            fits = false;
          }
          break;
        }
      }

      const clampedLeft = clamp(left, VIEWPORT_PADDING, viewportWidth - width - VIEWPORT_PADDING);
      const clampedTop = clamp(top, VIEWPORT_PADDING, viewportHeight - height - VIEWPORT_PADDING);

      if (candidate === 'top' || candidate === 'bottom') {
        arrowX = clamp(buttonCenterX - clampedLeft, MIN_ARROW_OFFSET, width - MIN_ARROW_OFFSET);
      } else {
        arrowY = clamp(buttonCenterY - clampedTop, MIN_ARROW_OFFSET, height - MIN_ARROW_OFFSET);
      }

      const nextPosition: PositionState = {
        top: clampedTop,
        left: clampedLeft,
        placement: candidate,
        arrowX,
        arrowY,
      };

      if (fits) {
        chosen = nextPosition;
        break;
      }

      if (!fallback) {
        fallback = nextPosition;
      }
    }

    if (!chosen) {
      chosen = fallback ?? {
        top: buttonRect.bottom + POPOVER_GAP,
        left: clamp(buttonCenterX - width / 2, VIEWPORT_PADDING, viewportWidth - width - VIEWPORT_PADDING),
        placement: placement,
        arrowX: clamp(width / 2, MIN_ARROW_OFFSET, width - MIN_ARROW_OFFSET),
        arrowY: clamp(height / 2, MIN_ARROW_OFFSET, height - MIN_ARROW_OFFSET),
      };
    }

    setPosition(chosen);
  }, [placement]);

  const closePopover = useCallback(
    (restoreFocus: boolean) => {
      shouldRestoreFocusRef.current = restoreFocus;
      if (restoreFocus) {
        suppressFocusOpenRef.current = true;
      }
      setIsOpen(false);
    },
    [],
  );

  const openPopover = useCallback(() => {
    shouldRestoreFocusRef.current = false;
    setIsOpen(true);
  }, []);

  useEffect(() => {
    if (!isOpen) {
      return;
    }

    const handlePointerDown = (event: PointerEvent) => {
      const target = event.target as Node | null;
      const button = buttonRef.current;
      const popover = popoverRef.current;
      if (!target) return;
      if (button && button.contains(target)) return;
      if (popover && popover.contains(target)) return;
      closePopover(false);
    };

    window.addEventListener('pointerdown', handlePointerDown);
    return () => {
      window.removeEventListener('pointerdown', handlePointerDown);
    };
  }, [isOpen, closePopover]);

  useEffect(() => {
    if (!isOpen) {
      return;
    }

    const popover = popoverRef.current;
    if (!popover) {
      return;
    }

    const focusInitial = () => {
      const focusable = getFocusableElements(popover);
      if (focusable.length > 0) {
        focusable[0].focus({ preventScroll: true });
      } else {
        popover.focus({ preventScroll: true });
      }
    };

    focusInitial();

    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        event.preventDefault();
        closePopover(true);
        return;
      }

      if (event.key === 'Tab') {
        const focusable = getFocusableElements(popover);
        if (focusable.length === 0) {
          event.preventDefault();
          popover.focus({ preventScroll: true });
          return;
        }

        const active = document.activeElement as HTMLElement | null;
        const first = focusable[0];
        const last = focusable[focusable.length - 1];

        if (event.shiftKey) {
          if (!active || active === first) {
            event.preventDefault();
            last.focus({ preventScroll: true });
          }
        } else if (!active || active === last) {
          event.preventDefault();
          first.focus({ preventScroll: true });
        }
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [isOpen, closePopover]);

  useEffect(() => {
    if (!isOpen) {
      return;
    }

    let frame = 0;
    const handle = () => {
      cancelAnimationFrame(frame);
      frame = requestAnimationFrame(() => {
        updatePosition();
      });
    };

    window.addEventListener('resize', handle);
    window.addEventListener('scroll', handle, true);

    return () => {
      cancelAnimationFrame(frame);
      window.removeEventListener('resize', handle);
      window.removeEventListener('scroll', handle, true);
    };
  }, [isOpen, updatePosition]);

  useIsomorphicLayoutEffect(() => {
    if (isOpen) {
      updatePosition();
    }
  }, [isOpen, updatePosition, tip.title, tip.bullets]);

  useEffect(() => {
    if (!isOpen) {
      if (shouldRestoreFocusRef.current) {
        shouldRestoreFocusRef.current = false;
        buttonRef.current?.focus({ preventScroll: true });
      }
    }
  }, [isOpen]);

  const handlePointerEnter = useCallback(
    (event: React.PointerEvent<HTMLButtonElement>) => {
      if (event.pointerType === 'mouse') {
        openPopover();
      }
    },
    [openPopover],
  );

  const handleFocus = useCallback(() => {
    if (suppressFocusOpenRef.current) {
      suppressFocusOpenRef.current = false;
      return;
    }
    if (!pointerDownRef.current) {
      openPopover();
    }
  }, [openPopover]);

  const handleClick = useCallback(() => {
    shouldRestoreFocusRef.current = false;
    setIsOpen(true);
  }, []);

  const handlePointerDown = useCallback(() => {
    pointerDownRef.current = true;
  }, []);

  const handlePointerUp = useCallback(() => {
    pointerDownRef.current = false;
  }, []);

  const popoverStyle = useMemo(() => {
    const style: CSSProperties & { '--arrow-x'?: string; '--arrow-y'?: string } = {
      top: `${position.top}px`,
      left: `${position.left}px`,
    };
    if (typeof position.arrowX === 'number') {
      style['--arrow-x'] = `${position.arrowX}px`;
    }
    if (typeof position.arrowY === 'number') {
      style['--arrow-y'] = `${position.arrowY}px`;
    }
    return style;
  }, [position]);

  return (
    <span className={combine('info-dot', className)}>
      <button
        ref={buttonRef}
        type="button"
        className="info-dot__button"
        aria-haspopup="dialog"
        aria-expanded={isOpen}
        aria-controls={popoverId}
        aria-label="Más información"
        title="Más información"
        data-state={isOpen ? 'open' : 'closed'}
        onClick={handleClick}
        onFocus={handleFocus}
        onPointerEnter={handlePointerEnter}
        onPointerDown={handlePointerDown}
        onPointerUp={handlePointerUp}
      >
        <span aria-hidden>i</span>
      </button>
      {isMounted && isOpen
        ? createPortal(
            <div
              ref={popoverRef}
              role="dialog"
              id={popoverId}
              aria-modal="false"
              aria-labelledby={titleId}
              tabIndex={-1}
              className="info-dot__popover"
              data-placement={position.placement}
              style={popoverStyle}
            >
              <h4 id={titleId} className="info-dot__title">
                {tip.title}
              </h4>
              <ul className="info-dot__list">
                {tip.bullets.map((bullet, index) => (
                  <li key={index}>{renderInline(bullet)}</li>
                ))}
              </ul>
            </div>,
            document.body,
          )
        : null}
    </span>
  );
}
