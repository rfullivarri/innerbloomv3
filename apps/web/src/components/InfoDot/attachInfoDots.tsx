import { createRoot, type Root } from 'react-dom/client';
import { InfoDot, type InfoDotPlacement } from './InfoDot';
import { infoTips, type InfoKey } from '../../content/infoTips';
import './InfoDot.css';

type DotEntry = {
  root: Root;
  mount: HTMLElement;
};

const registry = new WeakMap<HTMLElement, DotEntry>();

function isPlacement(value: unknown): value is InfoDotPlacement {
  return value === 'top' || value === 'bottom' || value === 'left' || value === 'right';
}

function isInfoKey(value: string | null | undefined): value is InfoKey {
  return Boolean(value && value in infoTips);
}

function collectHosts(target?: ParentNode): HTMLElement[] {
  const hosts: HTMLElement[] = [];

  if (typeof document === 'undefined') {
    return hosts;
  }

  if (!target) {
    document.querySelectorAll<HTMLElement>('[data-info]').forEach((element) => hosts.push(element));
    return hosts;
  }

  if (target instanceof HTMLElement && target.dataset.info) {
    hosts.push(target);
  }

  if ('querySelectorAll' in target) {
    target.querySelectorAll<HTMLElement>('[data-info]').forEach((element) => {
      if (!hosts.includes(element)) {
        hosts.push(element);
      }
    });
  }

  return hosts;
}

export function attachInfoDots(target?: ParentNode) {
  if (typeof window === 'undefined') {
    return;
  }

  const hosts = collectHosts(target);

  hosts.forEach((element) => {
    const key = element.dataset.info;
    if (!isInfoKey(key)) {
      return;
    }

    const placementAttr = element.dataset.infoPlacement ?? element.dataset.infoPos;
    const placement = isPlacement(placementAttr) ? placementAttr : undefined;

    const existing = registry.get(element);

    if (existing) {
      existing.root.render(<InfoDot id={key} placement={placement} />);
      return;
    }

    const mount = document.createElement('span');
    mount.className = 'info-dot__mount';

    if (placement === 'left') {
      element.insertBefore(mount, element.firstChild);
    } else {
      element.appendChild(mount);
    }

    const root = createRoot(mount);
    root.render(<InfoDot id={key} placement={placement} />);
    registry.set(element, { root, mount });
  });
}
