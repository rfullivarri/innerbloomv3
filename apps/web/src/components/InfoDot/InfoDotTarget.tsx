import { useEffect, useRef } from 'react';
import type { HTMLAttributes, PropsWithChildren } from 'react';
import { attachInfoDots } from './attachInfoDots';
import type { InfoKey } from '../../content/infoTips';
import type { InfoDotPlacement } from './InfoDot';
import './InfoDot.css';

interface InfoDotTargetProps extends PropsWithChildren, HTMLAttributes<HTMLDivElement> {
  id: InfoKey;
  placement?: InfoDotPlacement;
}

function combine(...classes: Array<string | false | null | undefined>) {
  return classes.filter(Boolean).join(' ');
}

export function InfoDotTarget({ id, placement, className, children, ...rest }: InfoDotTargetProps) {
  const hostRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const host = hostRef.current;
    if (!host) return;
    attachInfoDots(host);
  }, [id, placement]);

  return (
    <div
      ref={hostRef}
      data-info={id}
      data-info-placement={placement}
      className={combine('info-dot-target', className)}
      {...rest}
    >
      {children}
    </div>
  );
}
