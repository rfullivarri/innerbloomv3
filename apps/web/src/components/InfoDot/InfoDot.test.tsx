import { describe, expect, it } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import { act } from 'react';
import userEvent from '@testing-library/user-event';
import { InfoDot } from './InfoDot';

describe('InfoDot', () => {
  it('renders and opens on click', async () => {
    const user = userEvent.setup();
    render(<InfoDot id="xpLevel" />);

    const button = screen.getByRole('button', { name: /más información/i });
    await act(async () => {
      await user.click(button);
    });

    expect(await screen.findByRole('dialog')).toBeInTheDocument();
  });

  it('closes on outside click', async () => {
    const user = userEvent.setup();
    render(<InfoDot id="xpLevel" />);

    const button = screen.getByRole('button', { name: /más información/i });
    await act(async () => {
      await user.click(button);
    });
    await screen.findByRole('dialog');

    await act(async () => {
      await user.click(document.body);
    });
    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });

  it('closes on escape and restores focus', async () => {
    const user = userEvent.setup();
    render(<InfoDot id="xpLevel" />);

    const button = screen.getByRole('button', { name: /más información/i });
    await act(async () => {
      await user.click(button);
    });
    await screen.findByRole('dialog');

    await act(async () => {
    await act(async () => {
      await user.keyboard('{Escape}');
    });
    });
    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
    expect(document.activeElement).toBe(button);
  });

  it('matches the xpLevel snapshot', async () => {
    const user = userEvent.setup();
    render(<InfoDot id="xpLevel" />);
    await act(async () => {
      await user.click(screen.getByRole('button', { name: /más información/i }));
    });

    const dialog = await screen.findByRole('dialog');
    expect(dialog).toMatchSnapshot();
  });
});
