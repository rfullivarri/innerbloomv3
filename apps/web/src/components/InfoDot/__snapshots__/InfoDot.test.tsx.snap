// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InfoDot > matches the xpLevel snapshot 1`] = `
<div
  aria-labelledby=":r3:-title"
  aria-modal="false"
  class="info-dot__popover"
  data-placement="bottom"
  id=":r3:-dialog"
  role="dialog"
  style="top: 18px; left: 12px; --arrow-x: -11px;"
  tabindex="-1"
>
  <h4
    class="info-dot__title"
    id=":r3:-title"
  >
    ¿Qué ves acá?
  </h4>
  <ul
    class="info-dot__list"
  >
    <li>
      🏆 
      <strong>
        XP
      </strong>
       = experiencia total acumulada.
    </li>
    <li>
      🎯 
      <strong>
        Level
      </strong>
       sube al alcanzar el umbral de XP.
    </li>
    <li>
      📊 La barra muestra el 
      <strong>
        progreso al próximo nivel
      </strong>
      .
    </li>
    <li>
      💡 Cada hábito suma XP: pequeños pasos, gran avance.
    </li>
  </ul>
</div>
`;
